### PR TITLE
Add prometheus signer and cosigner metrics

### PR DESCRIPTION
**Pull request by chillyvee**
## Description

This PR addresses basic prometheus metrics mentioned in https://github.com/strangelove-ventures/horcrux/issues/11

This covers latest signed height/round, latest/missed heights, sentry dial failure counts.

## Checklist

- [X] I have made sure the upstream branch for this PR is correct
- [X] I have made sure this PR is ready to merge
- [X] I have made sure that I have assigned reviewers related to this project

## Changes

- [X] Added Basic Prometheus Metrics
- [X] Prevents Sentry boot crash loops with shorter dial timeouts.  Helpful when Sentry has long start times (full nodes)
- [X] Code upgrades for go 1.16 deprecations
- [X] Fixed formatting/lint issues for existing code

## Single Signer Testing

- Start Horcrux
- Wait
- curl http://localhost:2112/metrics
- Look at 'signer_total_sentry_connect_tries' increase
- Start Validator
- Wait
- Look at 'signer_last_precommit_height' increase
- Look at 'signer_last_prevote_height' increase
- Stop validator
- Look at 'signer_seconds_since_last_precommit' increase
- Look at 'signer_seconds_since_last_prevote' increase
- Wait 60 seconds
- Start validator
- Look at 'signer_missed_prevotes' increase
- Wait for validator to sync
- Look at 'signer_total_missed_prevotes' increase
- Look at 'signer_total_missed_precommits' increase

## Co-Signer Testing

- Look at docs/metrics.md to see key metrics to monitor
- Specifically 'signer_seconds_since_last_local_ephemeral_share_time' will indicate when a cosigner node is not contacted
- Stop 2 of 3 nodes and watch the last node signer_seconds_since_last_local_ephemeral_share_time increase

## Notes

Feedback welcome in general

Prefer to address any raft p2p specific / additional metrics on a separate commit/PR

This PR focuses on what can be seen outside the raft P2P layer

There are some lint failures with fixes in an isolated commit, but it is mostly related to existing leading spaces in boilerplate apache license comments or for comment clarity.

In addition, gofmt with -s or -w options on the flagged files does not produce any differences.

```
cmd/horcrux/main.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
cmd/horcrux/cmd/key2shares.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
cmd/horcrux/cmd/version.go:8: File is not `gofmt`-ed with `-s` (gofmt)
    http://www.apache.org/licenses/LICENSE-2.0
test/test_signer.go:120: File is not `gofmt`-ed with `-s` (gofmt)
//       same sentry node.
```

Removed the spaces in my last commit to reduce chances of human PR rejection, but I think it's more a "bug" or configuration issue in the linter.

### CODE REVIEW

**Comment by chillyvee** [comment]
@agouin - Want to try to run that lint check again?

**Comment by nitronit** [comment]
@chillyvee You can run the linter in your PR if you either add your bransch name here 
https://github.com/chillyvee/horcrux/blob/49cbbb2eccbed123031711f8012ef923de78b0e2/.github/workflows/lint.yml#L8
or
you push/merge your bransch to your fork main.

Note also that github actions always uses the latest release of linter in this case I think its 1.49. 
So to run the linter locally and fix the issue you need to run that version. If you are on mac for example and uses **brew** it doesnt install the latest version so you need to install the latest go-linter.

**Comment by chillyvee** [comment]
Thanks for the tips @nitronit !  I ran it locally (v1.49 as you stated)

Only a few messages were left (as noted above, related to existing code comments)

In general, very minor and seems removing the leading spaces actually makes readability "worse" so I decided not to fix it.

**Comment by chillyvee** [comment]
Whoops, didn't realize we created a conflict here.  Let's see if this latest edit passes all the tests.

**Comment by chillyvee** [comment]
A sample Grafana dashboard screenshot utilizing the Prometheus metrics is provided here:
https://github.com/chillyvee/horcrux-info

**Comment by agouin** [comment]
Thank you for your PR @chillyvee ! This is looking great. In order to get this merged in, we would like to handle #104 first so that we can have a single listener for the debug server (which doesn't exist yet in horcrux) and the prometheus metrics.

**Comment by chillyvee** [comment]
> Thank you for your PR @chillyvee ! This is looking great. In order to get this merged in, we would like to handle #104 first so that we can have a single listener for the debug server (which doesn't exist yet in horcrux) and the prometheus metrics.

Understand.  Do you already have a developer for the task?

**Comment by chillyvee** [comment]
Seeing a failed test, but due to the nature of the failure, maybe someone should take a look and see if it's a bug hiding somewhere, or a false alarm.

**Comment by chillyvee** [comment]
Guess it's just a intermittent failure in the test cases

File: cmd/horcrux/cmd/metrics.go
Comment (DavidNix) [inline cmd/horcrux/cmd/metrics.go:47]: UX nit: Logging an error indicates to the operator they did something wrong. But if they don't set the listen address, then that was purposeful, not an error. How about making this an `Info` instead?

We probably want to expand the log message "debug-listen-address not defined; debug server disabled"

File: cmd/horcrux/cmd/config.go
Comment (DavidNix) [inline cmd/horcrux/cmd/config.go:171]: I saw we enable the debug server by default. I would imagine most operators would want this functionality. So we make it opt out. A default will also show the expected string format: e.g. `:8543` or `localhost:8543`

File: docs/metrics.md
Comment (DavidNix) [inline docs/metrics.md:66]: s/Grapfana/Grafana

File: cmd/horcrux/cmd/metrics.go
Comment (DavidNix) [inline cmd/horcrux/cmd/metrics.go:29]: It's safe to unregister the sink at the conclusion of this function?

File: signer/metrics.go
Comment (DavidNix) [inline signer/metrics.go:18]: change request:

The time variables must be synchronized, otherwise there are likely data races. 

[go StartMetrics()](https://github.com/strangelove-ventures/horcrux/pull/93/files#diff-9ba01267c02440d5f34d850290066f2b246faf6d49ebc06ac3f1c8c42494531fR265) is reading values while another goroutine is writing to them.

A mutex may be appropriate but will probably litter the codebase with mu `Lock/Unlock`s.

**Review by DavidNix — CHANGES_REQUESTED** [review]
@chillyvee 

Thanks for your contribution! I'm a staff engineer at Strangrlove. I also appreciate the documentation!

Most of my comments are nits, but the data races with some package level variables promoted my change request.

**Comment by chillyvee** [comment]
@DavidNix 
Thank you for your thoughtful review.  Will get back to you with adjustments.

File: .github/workflows/tests.yml
Comment (chillyvee) [reply .github/workflows/tests.yml:8]: At first, getting CI tests to run prior to first merge here in this upstream repo was a little tricky (github blocks first time contributors from workflows).  So, it was useful to have a branch to push to in the fork for a lint/test.

Otherwise, there are some surprises for contributors when they send in a PR only to get lint-rejected days later.

This can be removed/reconfigured in many ways to achieve the same goal.  What do you think?

(meant to remove this and make it a separate issue, but since we're both looking at it, might as well ask :) )

**Comment by chillyvee** [comment]
Adjustments made.  Running through tests.

File: cmd/horcrux/cmd/config.go
Comment (chillyvee) [reply cmd/horcrux/cmd/config.go]: changed to debuglisten

File: cmd/horcrux/cmd/metrics.go
Comment (chillyvee) [reply cmd/horcrux/cmd/metrics.go]: removed if

File: cmd/horcrux/cmd/metrics.go
Comment (chillyvee) [reply cmd/horcrux/cmd/metrics.go]: implemented panic

File: cmd/horcrux/cmd/metrics.go
Comment (chillyvee) [reply cmd/horcrux/cmd/metrics.go]: implemented panic

File: cmd/horcrux/cmd/metrics.go
Comment (chillyvee) [reply cmd/horcrux/cmd/metrics.go:43]: Info with suggested message

File: cmd/horcrux/cmd/config.go
Comment (chillyvee) [reply cmd/horcrux/cmd/config.go]: Added example as suggested

File: docs/metrics.md
Comment (chillyvee) [reply docs/metrics.md]: fixed

File: cmd/horcrux/cmd/metrics.go
Comment (chillyvee) [reply cmd/horcrux/cmd/metrics.go]: Better not to unregister :)

File: signer/metrics.go
Comment (chillyvee) [reply signer/metrics.go:19]: Mutex implemented.

**Review by mark-rushakoff — COMMENTED** [review]
There are a couple themes to clean up here around passing contexts for clean shutdown and gradually moving away from global singletons.

There is enough subtlety in starting a background http server, that getting the debug server running with just the pprof endpoints, would be a detailed enough PR on its own, without tangling in Prometheus metrics wiring.

**Comment by chillyvee** [comment]
Thank you @mark-rushakoff  
We'll give a shot at adjusting the points you mentioned.

File: cmd/horcrux/cmd/config.go
Comment (chillyvee) [reply cmd/horcrux/cmd/config.go:125]: Yes!

File: cmd/horcrux/cmd/config.go
Comment (chillyvee) [reply cmd/horcrux/cmd/config.go:77]: Ok

File: cmd/horcrux/cmd/config.go
Comment (chillyvee) [reply cmd/horcrux/cmd/config.go:495]: Ok

File: cmd/horcrux/cmd/state.go
Comment (chillyvee) [reply cmd/horcrux/cmd/state.go:105]: No problem updating the line.
Did you want the rest of the file updated as well to match this pattern?
Let me know if errors should be output differently.

File: .github/workflows/tests.yml
Comment (DavidNix) [reply .github/workflows/tests.yml:8]: Thanks for the  explanation and we'll note the less than ideal user experience as a first time contributor. 

I think we can remove it.

File: cmd/horcrux/cmd/signer.go
Comment (chillyvee) [reply cmd/horcrux/cmd/signer.go]: Added cmd.Context() 

Also noticed that cmd/horcrux/cmd/root.go calls rootCmd.Execute() without any context being set.

File: cmd/horcrux/cmd/cosigner.go
Comment (chillyvee) [reply cmd/horcrux/cmd/cosigner.go]: Added cmd.Context()

Also noticed that cmd/horcrux/cmd/root.go calls rootCmd.Execute() without any context being set.

File: .github/workflows/tests.yml
Comment (chillyvee) [reply .github/workflows/tests.yml]: The relevant commit has been removed

**Comment by chillyvee** [comment]
CI tests look to be passing.   Let us know if any other changes needed.

File: signer/metrics.go
Comment (DavidNix) [inline signer/metrics.go:118]: Should these gauges be counters instead?

File: signer/metrics.go
Comment (DavidNix) [inline signer/metrics.go:158]: Should be counters?

File: cmd/horcrux/cmd/metrics.go
Comment (DavidNix) [inline cmd/horcrux/cmd/metrics.go:98]: This is just an FYI, not change request. 

When the process exits, nothing waits for these goroutines to finish, so the there's a good chance the debug server does not shut down cleanly. Given it's a debug server, I think it's ok to exit abruptly. We can refactor later.

**Review by DavidNix — APPROVED** [review]
@chillyvee Thanks again for your contribution. 

I have some nits, but otherwise looks good!

It's probably worth auditing the new metrics (there's quite a few of them) and doublecheck the metric type (counter vs. gauge) is set appropriately. 

I won't merge until @mark-rushakoff gives 👍 as well.

File: signer/metrics.go
Comment (chillyvee) [reply signer/metrics.go:118]: Good chance.  Let me check/adjust

File: cmd/horcrux/cmd/metrics.go
Comment (chillyvee) [reply cmd/horcrux/cmd/metrics.go:98]: Yes.  Probably needs to get a real context from the cmd.Execute call, and of course a resulting finish signal from the debug server.

Worst case here, a http body doesn't get to complete sending.  Safe to assume most modern clients will survive?

**Comment by chillyvee** [comment]
Changed Gauges to Counters for metric totals

**Comment by chillyvee** [comment]
This CI test seems to succeed/fail a little randomly as commits get added to the same PR.  Is it known issue/memory shortage/actual problem?

```
=== CONT  Test2Of3SignerTwoSentries
    horcrux_test.go:74: 
        	Error Trace:	horcrux_test.go:74
        	Error:      	Received unexpected error:
        	            	All attempts fail:
        	            	#1: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#2: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#3: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#4: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#5: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#6: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#7: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#8: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#9: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	            	#10: post failed: Post "http://localhost:55696": dial tcp 127.0.0.1:55696: connect: connection refused
        	Test:       	Test2Of3SignerTwoSentries
```

File: cmd/horcrux/cmd/metrics.go
Comment (mark-rushakoff) [inline cmd/horcrux/cmd/metrics.go:66]: Commented line can be removed.

File: cmd/horcrux/cmd/metrics.go
Comment (mark-rushakoff) [inline cmd/horcrux/cmd/metrics.go:78]: Better to check `errors.Is(http.ErrServerClosed)` rather than string comparison -- the Go team has made changes to string error content in the past.

**Review by mark-rushakoff — APPROVED** [review]
I left a couple small notes, but I think this can be merged now.

File: cmd/horcrux/cmd/metrics.go
Comment (chillyvee) [reply cmd/horcrux/cmd/metrics.go]: Removed line

File: cmd/horcrux/cmd/metrics.go
Comment (chillyvee) [reply cmd/horcrux/cmd/metrics.go]: Changed to errors.Is()

**Comment by chillyvee** [comment]
Thank you @nitronit @agouin @DavidNix @bussyjd @mark-rushakoff for your feedback

**Pull request merged**
